reference/src/layout/structs-and-tuples.md

```diff
@@ -122,6 +122,27 @@ compiler will not reorder it, to allow for the possibility of
 unsizing. E.g., `struct Foo { x: u16, y: u32 }` and `struct Foo<T> {
 x: u16, y: T }` where `T = u32` are not guaranteed to be identical.
 
+#### Zero-sized structs
+
+Structs with default layout (`repr(Rust)`), layout with increased alignment
+(`repr(align(N))`), packed layout (`repr(packed(N))`), or C-compatible layout
+(`repr(C)`) are zero-sized, if they contain no fields of non-zero size. That is,
```

Comment:
Is this an "if and only if"?

Reply:
I think so. If a struct contains a field of non-zero size, I don't see how it could end up being zero-sized.

Reply:
I can't think of a counter-example, but it seems like only one direction is much commonly thought about and discussed, so I'd be a bit scared to commit to the other direction as well. Besides, it's usually more important to know when something definitely is a ZST (e.g., so you can ignore it for layout related questions), not when something definitely isn't a ZST.

Reply:
Fair, but but then we should try to word this in a way that it is clear that only one direction is defined. In particular with the way that conditional and conclusion are reversed here, I think it is easy to interpret this as bidirectional.

English is such a bad language for precise specifications...

Reply:
So, to make a concrete proposal (also removing a pointless double-negation):

> If all fields of a struct have size 0, then the struct has size 0.

Your sentence lists a few representations explicitly, but is there any repr for which this does *not* make sense?

Reply:
Well, that statement is trivially true if there struct has no fields.  C++ allows structs with no fields but gives them a size of 1, and in theory there could be some kind of `#[repr(C++)]` that accounts for this.

I'm not sure how plausible `#[repr(C++)]` is, but another hypothetical use case could be to fix the [ABI issue with passing small C++ classes by vaue](https://github.com/rust-lang/rust-bindgen/issues/778).

Reply:
> but is there any repr for which this does not make sense?

If we ever add a `repr()` for which this does not make sense, then this wouldn't hold for  `repr(transparent)` either.

The current wording mentions all `repr`s that currently exist, except for `repr(transparent)` , for which this just follows from its rules.

Reply:
If you really want to list the `repr`, I'd propose prepending something like:

> For `repr(Rust)`, `repr(packed(N))`, `repr(align(N))` and `repr(C)`:

(Btw, why is it `packed` but not `aligned`? Seems inconsistent...)

Reply:
I've reworded the guarantee with this feedback. Let me know if its ok now.